### PR TITLE
fix: Fixed artifact retrieval when templateRef in use. Fixes #9631, #9644.

### LINKED ISSUE #9631: Can't access artifacts via UI when artifacts come from template referenced by templateref
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

I expect to be able to download/view artifacts via the server-UI. This is only possible with artifacts which originate from templates that are defined within the same workflow/workflowtemplate.

### Version

3.4.0

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

Working Example:
```YAML
# This example demonstrates the ability to pass artifacts
# from one step to the next.
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: artifact-passing-
spec:
  serviceAccountName: argo-workflow
  entrypoint: artifact-example
  templates:
  - name: artifact-example
    steps:
    - - name: generate-artifact
        template: whalesay
    - - name: consume-artifact
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"

  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["sleep 1; cowsay hello world | tee /tmp/hello_world.txt"]
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt

  - name: print-message
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["cat /tmp/message"]
```

Failing example:
```YAML
# This example demonstrates the ability to pass artifacts
# from one step to the next.
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: artifact-passing-
spec:
  serviceAccountName: argo-workflow
  entrypoint: artifact-example
  templates:
  - name: artifact-example
    steps:
    - - name: generate-artifact
        templateRef:
          name: whalesay-template
          template: whalesay
    - - name: consume-artifact
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"

  - name: print-message
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["cat /tmp/message"]

---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: whalesay-template
spec:
  templates:
  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["sleep 1; cowsay hello world | tee /tmp/hello_world.txt"]
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt
```

### Logs from the server
```
time="2022-09-20T12:32:16.527Z" level=info msg="Get artifact file" artifactName=output namespace=argo nodeId=test-build-n5rw5-3520357635 workflowName=test-build-n5rw5
time="2022-09-20T12:32:16.542Z" level=error msg="Artifact Server returned internal error" error="no template found by the name of '' (which is the template associated with nodeId 'test-build-n5rw5-3520357635'??"
time="2022-09-20T12:32:16.542Z" level=info duration=15.435175ms method=GET path=/artifact-files/argo/workflows/test-build-n5rw5/test-n5rw5-3520357635/outputs/output size=22 status=500
time="2022-09-20T12:32:16.590Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetWorkflow grpc.service=workflow.WorkflowService grpc.start_time="2022-09-20T12:32:16Z" grpc.time_ms=9.277 span.kind=server system=grpc
time="2022-09-20T12:32:16.591Z" level=info duration=10.740922ms method=GET path=/api/v1/workflows/argo/demo-ascet-vecu-build-n5rw5 size=19328 status=0
time="2022-09-20T12:32:17.576Z" level=info duration=3m53.079937295s method=GET path=/api/v1/workflow-events/argo size=19516 status=0
```

Reference: [Discussion about this in Slack](https://cloud-native.slack.com/archives/C01QW9QSSSK/p1663676822966269?thread_ts=1663662972.187929&cid=C01QW9QSSSK)
[Examples.zip](https://github.com/argoproj/argo-workflows/files/9608357/Examples.zip)

### PR DESCRIPTION
**Pull request by brianloss**
Handle artifact retrieval when the workflow node has no templateName set but
does have templateRef set instead. Added a unit test that was broken before
the changes and works after the changes.

Fixes #9631, #9644
Signed-off-by: Brian Loss <brianloss@gmail.com>

### CODE REVIEW

**Comment by brianloss** [comment]
@juliev0 I saw you self-assigned #9631. Here is my attempt to fix it...

File: server/artifacts/artifact_server.go [hunk 1]
```diff
@@ -391,6 +391,10 @@ func (a *ArtifactServer) getArtifactAndDriver(ctx context.Context, nodeId, artif
 	// 4. Template defines ArchiveLocation
 
 	templateName := wf.Status.Nodes[nodeId].TemplateName
+	templateRef := wf.Status.Nodes[nodeId].TemplateRef
+	if templateName == "" && templateRef != nil {
+		templateName = templateRef.Template
+	}
```
Comment (juliev0) [inline server/artifacts/artifact_server.go:397]: thanks for fixing this! One thing I noticed is that in workflow/util/util.go there's a method called getTemplateFromNode() which is basically performing this logic. Maybe you can call that?

Comment (juliev0) [reply server/artifacts/artifact_server.go:397]: Another thing is that I somehow produced a scenario in which neither templateName nor templateRef is defined for a node (I'm referencing a ClusterWorkfowTemplate). Maybe I can provide you with the Workflow spec and ClusterWorkflowTemplate for that and we can make sure to handle that as best we can?

Comment (brianloss) [reply server/artifacts/artifact_server.go]: Sure, you can pass it along!

**Review by juliev0 — APPROVED** [review]

**Review by terrytangyuan — APPROVED** [review]

**Comment by brianloss** [comment]
@juliev0 or @terrytangyuan, do I need to do anything else to help get this fix gets into the next 3.4.x release? I wasn't sure since this went to master and there's a release-3.4 branch.

**Comment by terrytangyuan** [comment]
@sarabala1979 Could we cherry-pick this?

**Pull request merged**